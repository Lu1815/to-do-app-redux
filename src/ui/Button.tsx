import clsx from "clsx"

interface IButtonProps {
    onClick: () => void,
    text?: string,
    textColor?: string,
    fontWeight?: string,
    backgroundColor?: string,
    className?: string,
}

export const Button = ({ onClick, text, textColor, fontWeight, backgroundColor, className }: IButtonProps) => {
    
    const handleClick = () => {
        onClick();
    }

    return (
        <button className={
            clsx(
                className,
                'rounded-md px-3 py-2',
                (backgroundColor ?? 'bg-btn'),
                (textColor ?? 'text-white'),
                (fontWeight ?? 'font-normal')
            )}
            onClick={handleClick}
        >
            {text ?? 'Button'}
        </button>
    )
}