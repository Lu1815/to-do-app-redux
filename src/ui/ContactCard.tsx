interface IContactCardProps {
    contactName: string,
    contactImage: string
}

const ContactCard = ({ contactImage, contactName }: IContactCardProps) => {
    return (
        <div className="bg-white border-gray-200 border-2 py-8 px-10 text-center rounded-md shadow-lg w-full mx-auto">
            <div className="flex flex-row gap-3 justify-left mx-auto">
                <img className="w-20 h-20 object-cover rounded-full shadow-lg" src={contactImage} alt="Contact Avatar" />
                <div className="grid content-center">
                    <h2 className={"font-semibold text-left align-middle max-h-min text-2xl md:text-lg lg:text-2xl break-words"}>{contactName}</h2>
                </div>
            </div>
        </div>
    )
}

export default ContactCard