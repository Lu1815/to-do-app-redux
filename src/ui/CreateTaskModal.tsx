import { Dispatch, SetStateAction, useState } from "react";
import Input from "./Input";
import { useAppDispatch, useAppSelector } from "../App/redux/Hooks";
import { ITodoState, create } from "../App/redux/TodosSlice";
import { ToastContainer, toast } from "react-toastify";
import { ErrorNotification } from "../lib/Notifications";

const DEFAULT_INT = 0;

interface ICreateCardModalProps {
    showModal: boolean;
    setShowModal: Dispatch<SetStateAction<boolean>>;
}

const CreateTaskModal = ({ showModal, setShowModal }: ICreateCardModalProps) => {
    const todos: ITodoState[] = useAppSelector((state) => state.todos)
    const dispatch = useAppDispatch()
    const [taskName, setTaskName] = useState<string>("");
    const [error, setError] = useState<boolean>(false);
    const inputProps = { taskName, setTaskName, error, setError };

    const handleHideModal = () => {
        setShowModal(false)
    }

    const handleAddTask = () => {
        if (!taskName) {
            ErrorNotification('Task name cannot be empty!')
            setError(true);
            return;
        }

        let todoId: number = DEFAULT_INT;

        if(todos.length <= 0){
            todoId = 1;
        } else {
            todoId = todos[todos.length - 1].id + 1;
        }

        const todoItem: ITodoState = {
            id: todoId,
            name: taskName,
        }

        dispatch(create(todoItem))
        setTaskName("");
        setShowModal(false)
    }

    return (
        <>
            {
                showModal && (
                    <div className="bg-black bg-opacity-50 h-svh w-max">
                        <div id="popup-modal" tabIndex={-1} data-modal-backdrop="static" className="fixed inset-0 flex items-center justify-center h-screen w-screen bg-black bg-opacity-50">
                            <div className="relative p-4 w-full max-w-md max-h-full">
                                <div className="relative bg-white rounded-lg shadow">
                                    <div className="p-4 md:p-5 text-center flex flex-col">
                                        <h3 className="mb-5 text-lg font-normal text-gray-500 dark:text-gray-400">Task Name:</h3>
                                        <Input {...inputProps} />
                                        <div>
                                            <button onClick={handleHideModal} type="button" className="text-white bg-red-600 hover:bg-red-800 focus:ring-4 focus:outline-none focus:ring-red-300 dark:focus:ring-red-800 font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center me-2 mt-2">
                                                Cancel
                                            </button>
                                            <button onClick={handleAddTask} type="button" className="bg-btn hover:bg-btn-hovered focus:ring-4 focus:outline-none focus:ring-gray-200 rounded-lg border border-gray-200 text-sm font-medium px-5 py-2.5 focus:z-10">
                                                Add Task
                                            </button>

                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                )
            }
            <ToastContainer />
        </>
    )
}

export default CreateTaskModal