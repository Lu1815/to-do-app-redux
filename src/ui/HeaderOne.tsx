import clsx from 'clsx'

interface IHeaderOneProps {
    content?: string,
    color?: string,
    weigth?: string
}

const HeaderOne = ({ content, color, weigth }: IHeaderOneProps) => {
  return (
    <p className={
        clsx(
            'text-lg',
            (color ?? 'text-black'),
            (weigth ?? 'font-normal')
        )
    }>
        { content ?? 'Normal Text' }
    </p>
  )
}

export default HeaderOne