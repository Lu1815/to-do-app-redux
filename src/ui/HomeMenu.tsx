import { Link } from "react-router-dom"

export const HomeMenu = () => {
    return (
        <div className="container h-dvh mx-auto grid content-center">
            <div className="flex flex-col justify-center gap-2 text-center mx-auto w-2/3">
                <Link to="/tasks" className="bg-btn rounded-md text-white font-bold px-3 py-2">
                    Tasks
                </Link>
                <Link to="/list" className="bg-btn rounded-md text-white font-bold px-3 py-2">
                    List
                </Link>
            </div>
        </div>
    )
}