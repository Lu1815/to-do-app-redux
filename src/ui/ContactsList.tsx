import { useGetElementsListQuery } from "../App/redux/MockAPI"
import ContactCard from "./ContactCard"
import Loader from "./Loader";

const ContactsList = () => {
    const { data: contactList, isLoading } = useGetElementsListQuery();

    if (isLoading) return <Loader />

    return (
        <div className='flex flex-col py-3 gap-3'>
            No image loads, because the images URLs are unreachable
            <div className="flex flex-col md:grid md:grid-cols-3 lg:grid-cols-4 justify-center gap-2 py-3">
                {
                    contactList!.length === 0 ? <h1>There are no contacts yet...</h1>
                        :
                        contactList!.map((contact) => (
                            <ContactCard contactName={contact.name} contactImage={contact.avatar} key={contact.id} />
                        ))
                }
            </div>
        </div>
    )
}

export default ContactsList