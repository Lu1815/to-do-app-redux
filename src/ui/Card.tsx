import React from 'react'
import HeaderOne from './HeaderOne'

interface ICardProps {
    text?: string
}

const Card = ({text}: ICardProps) => {
  return (
    <div className='w-full py-2 px-3 border-gray-200 border-2 rounded-md'>
        <HeaderOne content={text ?? 'Card Text'} />
    </div>
  )
}

export default Card