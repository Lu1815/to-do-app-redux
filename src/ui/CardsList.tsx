import Card from './Card'
import { ITodoState } from '../App/redux/Index'

interface ICardsListProps {
    todosList: ITodoState[],
}

const CardsList = ({todosList}: ICardsListProps) => {
  return (
    <div className='flex flex-col justify-center gap-2 py-3'>
        {
          todosList.length === 0 ? <h1>There are no tasks yet...</h1>
          :
            todosList.map((todoItem) => (
                <Card text={todoItem.name} key={todoItem.id}/>
            ))
        }
    </div>
  )
}

export default CardsList