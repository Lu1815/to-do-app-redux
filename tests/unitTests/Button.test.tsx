import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Button } from '../../src/ui/Button';

describe('Button Component', () => {
  it('renders without crashing', () => {
    render(<Button onClick={() => {}} />);
    expect(screen.getByText('Button')).toBeInTheDocument();
  });

  it('renders with custom text', () => {
    render(<Button onClick={() => {}} text="Custom Button" />);
    expect(screen.getByText('Custom Button')).toBeInTheDocument();
  });

  it('executes onClick callback when clicked', () => {
    const mockClick = jest.fn();
    render(<Button onClick={mockClick} />);
    
    const button = screen.getByText('Button');
    fireEvent.click(button);
    
    expect(mockClick).toHaveBeenCalledTimes(1);
  });

  it('applies custom text color', () => {
    render(<Button onClick={() => {}} textColor="text-red-500" />);
    const button = screen.getByText('Button');
    expect(button).toHaveClass('text-red-500');
  });

  it('applies custom background color', () => {
    render(<Button onClick={() => {}} backgroundColor="bg-blue-500" />);
    const button = screen.getByText('Button');
    expect(button).toHaveClass('bg-blue-500');
  });

  it('applies custom className', () => {
    render(<Button onClick={() => {}} className="custom-class" />);
    const button = screen.getByText('Button');
    expect(button).toHaveClass('custom-class');
  });

  it('matches snapshot', () => {
    const { container } = render(<Button onClick={() => {}} />);
    expect(container).toMatchSnapshot();
  });
});
