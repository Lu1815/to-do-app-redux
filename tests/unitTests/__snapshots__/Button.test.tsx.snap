// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<div>
  <button
    class="rounded-md px-3 py-2 bg-btn text-white font-normal"
  >
    Button
  </button>
</div>
`;
